Shopify/vscode-ruby-lsp
# Use testItem kind

### Motivation

https://github.com/Shopify/ruby-lsp/pull/734 added a `kind` to test items, of either `example` or `group`. The extension can use this to distinguish running a single tests from a full test file.

Previously we were doing this by checking for `test_` at the beginning of the name. But that approach isn't viable for some test frameworks, e.g. Rails declarative tests, or RSpec, which other Ruby LSP extensions may need to support.

### Implementation

Since we don't own the TestItem interface, there are limited way to supplement it with this additional information. I went with use a `TestTag`.

### Automated Tests

Nothing yet, it could be tricky. I'd like to collect some opinions on how to approach this.

### Manual Tests

Note that this relies on behaviour which in `main` of Ruby LSP but not yet released.

To ensure the new tag is being read, and it isn't just falling back to the previous behaviour, I suggest temporarily changing the two `test_` checks to `notest_` in `testController.ts`.

* Start debugging this branch, and open `ruby-lsp`.
* Open a test file contains multiple tests, e.g. `test/rubocop/cop/ruby_lsp/use_language_server_aliases_test.rb`.
* Click on the class 'Run in Terminal' and verify it runs the whole test on the command line.
* Click on the class 'Run' and verify it runs the whole test on the command line.
* Click on the individual test 'Run' and verify is run in the Test Explorer.
* Set a breakpoint by clicking the gutter. Click 'Debug' and ensure it pauses.
